packages/runtime/container-runtime/src/containerRuntime.ts

```diff
@@ -2407,83 +2429,63 @@ export class ContainerRuntime
 				0x791 /* Connect event delay must be canceled before subsequent connect event */,
 			);
 			assert(!!clientId, 0x792 /* Must have clientId when connecting */);
+
```

Comment:
we should just remove this whole block, and move this.blobManager.hasPendingStashedBlobs() to canSendOps. i'm pretty sure we can just get rid of delayConnectClientId if we do that

Reply:
That's a good point, let me try that...

Reply:
@anthony-murphy , I tried, and I'm running into issues. Or rather, discovering issues.

I've added assertion that when `blobManager.processStashedChanges()` completes, `blobManager.hasPendingStashedBlobs()` should be false. That is not true, as `processStashedChanges()` completes when blobs are uploaded. But nobody is clearing `PendingBlob.pendingStashed`, because we are not sending blob ops for this blob! `opsent` is set to true for `pendingStashed` = true, which looks wrong, as it breaks the contract.

I changed `opsent`, and now I'm getting asserts in `processPendingLocalMessage()` as it does not expect that op - it was not in its list (or rather - it's probably at the end of the stack).

I can probably carve a solution that does not try to change all that, but the area looks a bit broken to me...

Reply:
Maybe to add - my assumption was - we can evaluate this.blobManager.processStashedChanges() promise only once (in constructor of ContainerRuntime). Existing code will go through this dance with promise and continuation on every reconnect, as this.blobManager.hasPendingStashedBlobs() would never switch to false. That seems like should have been called in UTs.

Reply:
I've published latest iteration with my attempt to carefully walk around these problems.
I'm using boolean (this.blobUploadCallbackInFlight) to track properly transitions, instead of relying on blobManager.processStashedChanges().

All test changes are borrowed from https://github.com/microsoft/FluidFramework/pull/20215 and will not be in this PR once above mentioned PR merges - they are required to satisfy requirement that clientID !== undefined on connected events.

Reply:
+ @dannimad

Reply:
Opened https://dev.azure.com/fluidframework/internal/_queries/edit/7617/?triage=true to track it.

Reply:
opsent is sent to true because we indeed sent an op while we were stashing the pending blob. Those pending stashed blobs should be clean once the op sent while stashing is processed and that's why we set opsent as true, otherwise when applying the pending state, we will complete the blob upload and send a new op. hasPendingStashedBlobs won't be false until stashed BlobAttach ops are processed so the added assertion makes a wrong assumption imo

Reply:
Let's continue discussion in the ADO item